feat(purchase): update purchase orders

jniles opened the pull request:
This commit implements the ability to update non-confirmed purchase orders.  If a purchase order has been confirmed, the only thing that should be updated is the status (canceled).  This is yet to be discussed.

To modify a purchase order, one must go to the purchase registry and select "edit" on the purchase orders that is non yet confirmed.  This will take you to the regular purchase order creation page with all values filled out.

Closes #5489.

Here is what it looks like:

![sXzDxvsomU](https://user-images.githubusercontent.com/896472/111638135-599d8000-87fa-11eb-9864-128e0c10e437.gif)


Note that this implementation breaks updating the status.  I'd like some help figuring out which status we should allow.  I'll fix that before a merge.

jniles committed:
feat(purchase): update purchase orders

This commit implements the ability to update non-confirmed purchase
orders.  If a purchase order has been confirmed, the only thing that
should be updated is the status (canceled).  This is yet to be
discussed.

To modify a purchase order, one must go to the purchase registry and
select "edit" on the purchase orders that is non yet confirmed.  This
will take you to the regular purchase order creation page with all
values filled out.

Closes #5489.

jmcameron commented on client/src/js/services/Store.js:
```diff
@@ -76,19 +74,19 @@ function StoreService() {
-    var data = this.data;
-    var index = this.index;
-    var identifier = this.identifier;
+    const { data } = this;
```
Can't these be combined?

jmcameron reviewed (commented):
I skimmed the code.  Hopefully @mbayopanda will take a closer look at the code than I did.

I tested it with bhima_test.  It works and is a useful addition.

I ran into a couple of minor non-blocking issues:
- The edited form will not allow submission until notes are added.   Why is the description required?    If it is required, we should update the Bhima_test SQL files to add a non-null note to the unfulfilled purchase requisition (which is dated in 2016).   If this gets modified, you may as well add a relative date on it.

- The purchase order report has a minor/trivial typo.  The 's' of Note(S) is a capital letter.  It should be lower case:
![image](https://user-images.githubusercontent.com/62145/111698659-51424680-87f4-11eb-9e86-44709e86c5e9.png)

jniles committed:
fix(purchase): allow canceling orders

This commit allows purchase orders to be canceled.  It also adds a
updated_at column to the purchase table so that we can keep track of
when edits happen.  Finally, canceling a purchase order is a sink - we
can't undo a cancel; however, you can still cancel an order that is
confirmed.

Closes #5489.

jmcameron commented on test/data.sql:
```diff
@@ -590,17 +590,18 @@ INSERT INTO `price_list_item` VALUES
-INSERT INTO `purchase` VALUES
-  (@purchase_order, 1, 1, 300, 2, HUID('3ac4e83c-65f2-45a1-8357-8b025003d793'), DATE_ADD(CURRENT_DATE, INTERVAL -1725 DAY), CURRENT_TIMESTAMP, 1, NULL, NULL, 1);
+INSERT INTO `purchase`
+  (uuid, project_id, reference, cost, currency_id, supplier_uuid, date, user_id, payment_method, status_id) VALUES
```
I suggest adding a note here (like the one below).

jmcameron reviewed (approved):
Looks good to me.  I had one question about test/data.sql but there may be a reason for leaving out the 'note' field on some of the purchases.  In any case we can deal with it later.

I ran all tests on my local PC and all but E2E ran fine.  (The issue with my failing E2E test is not related to this issue; it fails on master too).

mbayopanda reviewed (commented):
When we create a purchase order it's already confirmed, so we cannot test the edit feature

mbayopanda commented on client/src/modules/purchases/modals/status.tmpl.html:
```diff
@@ -5,32 +5,32 @@
+<form
   name="ModalForm"
-  ng-submit="$ctrl.submit($ctrl.status)"
+  bh-submit="$ctrl.submit($ctrl.status)"
```
👍🏽

mbayopanda commented on server/controllers/finance/purchases.js:
```diff
@@ -299,23 +301,78 @@ function detail(req, res, next) {
+    // can't edit non-awaiting-confirmation records
+    if (!isCancelRequest && purchase.status_id !== PURCHASE_STATUS_WAITING_CONFIRMATION) {
+      throw new BadRequest('Can only modify purchase orders that are awaiting confirmation.');
+    }
```
👍🏽 these checks on the server-side are excellent

mbayopanda commented on server/controllers/finance/purchases.js:
```diff
@@ -299,23 +301,78 @@ function detail(req, res, next) {
+    await txn.execute();
+
+    // we only need the second return value
+    const [, record] = await Promise.all([
```
`[,` so sexy 😆

mbayopanda reviewed (commented):
The edit purchase feature works fine 👍🏽 the only issue I got is when we create a new purchase order it is already confirmed

mbayopanda reviewed (approved):
LGTM 👍🏽
